### PR TITLE
oidc authenticator: allow string value as groups claim

### LINKED ISSUE #33290: OpenIDConnect - group claim assumes list of strings, should accept single value
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

No, already discussed on sig-auth

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

confirmed issue on sig-auth channel

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

Depends on interpretation

**Kubernetes version** (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.0", GitCommit:"283137936a498aed572ee22af6774b6fb6e9fd94", GitTreeState:"clean", BuildDate:"2016-07-01T19:26:38Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.4+coreos.0", GitCommit:"be9bf3e842a90537e48361aded2872e389e902e7", GitTreeState:"clean", BuildDate:"2016-08-02T00:54:53Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**:  CoreOS Vagrant Image
- **OS** (e.g. from /etc/os-release): 

NAME=CoreOS
ID=coreos
VERSION=1122.0.0
VERSION_ID=1122.0.0
BUILD_ID=2016-07-27-0739
PRETTY_NAME="CoreOS 1122.0.0 (MoreOS)"
ANSI_COLOR="1;32"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"

- **Kernel** (e.g. `uname -a`):

Linux localhost 4.7.0-coreos #1 SMP Wed Jul 27 07:30:04 UTC 2016 x86_64 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz GenuineIntel GNU/Linux

- **Install tools**:

Vagrant

- **Others**:

**What happened**:

The claim named in --oidc-groups-claim MUST be list of strings, even if there is only one value.  Talking to @ericchiang this is by design since its assumed that users will have more then one group.

**What you expected to happen**:

The OIDC authenticator should accept either a list of strings or a single string since:

1.  There is no concept of schema that marks an attribute as single value or multi value in JSON or a JWT
2.  There is no standard for representing groups in a JWT
3.  There is no standard for representing groups in OIDC

An OIDC should be flexible in how it accepts data.

**How to reproduce it** (as minimally and precisely as possible):

1.  Deploy with RBAC enabled
2.  Create a role binding for the group "admin"
3.  Generate a JWT such as the one that does NOT work below
4.  User won't be able to login
5.  Generate a JWT where the role is multi-value
6.  User will be authorized

Does *NOT* Work:
```json
{
  "iss": "https://mlb.tremolo.lan:8043/auth/idp/oidc",
  "aud": "kubernetes",
  "exp": 1474562249,
  "jti": "8UgF8is2XOgqPUlExxlsdg",
  "iat": 1474562189,
  "nbf": 1474562069,
  "sub": "mmosley",
  "user_role": "admin",
  "email": "matt.mosley@tremolosecurity-test.com"
}
```

*DOES* Work:
```json
{
  "iss": "https://mlb.tremolo.lan:8043/auth/idp/oidc",
  "aud": "kubernetes",
  "exp": 1474564875,
  "jti": "3Oo35Su-oXPX3RoUY91xXg",
  "iat": 1474564815,
  "nbf": 1474564695,
  "sub": "mmosley",
  "user_role": [
    "admin",
    "users",
    "approvers"
  ],
  "email": "matt.mosley@tremolosecurity-test.com"
}
```

**Anything else do we need to know**:

### PR DESCRIPTION
**Pull request by ericchiang**
Allow the group claim to be a single string instead of an array of
strings. This means the following claim

```
{
  "role": "admin"
}
```

Will be mapped to the groups

```
["admin"]
```

cc @kubernetes/sig-auth @mlbiam

closes #33290

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33332)

### CODE REVIEW

File: pkg/genericapiserver/options/server_run_options.go [hunk 1]
```diff
@@ -401,7 +401,7 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
 
 	fs.StringVar(&s.OIDCGroupsClaim, "oidc-groups-claim", "", ""+
 		"If provided, the name of a custom OpenID Connect claim for specifying user groups. "+
-		"The claim value is expected to be an array of strings. This flag is experimental, "+
+		"The claim value is expected to be a string or array of strings. This flag is experimental, "+
```
Comment (lavalamp) [inline pkg/genericapiserver/options/server_run_options.go]: This seems like an unnecessary change; what is wrong with a 1-element list? How is that different?

Comment (liggitt) [inline pkg/genericapiserver/options/server_run_options.go]: Scalar string, not a one element list

**Comment by lavalamp** [comment]
right, my question is, why is this distinction important?

File: pkg/genericapiserver/options/server_run_options.go (hunk 1 above)
Comment (liggitt) [inline pkg/genericapiserver/options/server_run_options.go]: And the claim structure is frequently outside the control of the relying party, so "just change the claim structure returned by the IDP" is often not an option

**Comment by liggitt** [comment]
To expand the number and type of claims that could be used to convey group membership

**Comment by lavalamp** [comment]
lgtm
